srml/contract/src/lib.rs

```diff
@@ -14,41 +14,74 @@
 // You should have received a copy of the GNU General Public License
 // along with Substrate. If not, see <http://www.gnu.org/licenses/>.
 
-//! Smart-contract module for runtime; Allows deployment and execution of smart-contracts
-//! expressed in WebAssembly.
+//! # Contract Module
+//! 
+//! The contract module provides functionality for the runtime to deploy and execute WebAssembly smart-contracts.
+//! The supported dispatchable functions are documented as part of the [`Call`](./enum.Call.html) enum.
+//! 
+//! ## Overview
+//! 
+//! This module extends accounts (see `Balances` module) to have smart-contract functionality.
+//! These "smart-contract accounts" have the ability to create smart-contracts and make calls to other contract and non-contract accounts.
+//! 
+//! The smart-contract code is stored once in a code cache, and later retrievable via a code hash.
+//! This means that multiple smart-contracts can be instantiated from the same code cache, without replicating the code each time.
+//! 
+//! When a smart-contract is called, its associated code is retrieved via the code hash and gets executed.
+//! This call can alter the storage entries of the smart-contract account, create new smart-contracts, or call other smart-contracts.
+//! 
+//! Finally, when the `balances` module determines an account is dead (i.e. account balance fell below the
+//! existential deposit), it reaps the account. This will delete the associated code and storage of the smart-contract account.
+//! 
+//! **NOTE:** Call failures are not always cascading. For example, if contract A calls contract B, and B
+//! fails, A can decide how to handle that failure, either proceeding or reverting A's changes.
+//! 
+//! ### Gas
+//! 
+//! Senders must specify a gas limit with every call, as all instructions invoked by the smart-contract require gas.
+//! Unused gas is refunded after the call, regardless of the execution outcome.
+//! 
+//! If the gas limit is reached, then all calls and state changes (including balance transfers) are only reverted at the current call's contract level.
+//! For example, if contract A calls B and B runs out of gas mid-call, then all of B's calls are reverted.
+//! Assuming correct error handling by contract A, A's other calls and state changes still persist. 
+//! 
+//! ## Interface
+//! 
+//! ### Dispatchable functions
+//! 
+//! * `put_code` - Stores the given binary Wasm code into the chains storage and returns its `codehash`.
+//! 
+//! * `create` - Deploys a new contract from the given `codehash`, optionally transferring some balance. This creates a new smart contract account and calls its contract deploy handler to initialize the contract.
```

Comment:
In our style guide we limit the code width to be around 100 columns. Can we wrap this line?